File: day19/main.go
package main

import (
	_ "embed"
	"log/slog"
	"strings"
	"time"
)

//go:embed inp
var inp string

func main() {
	slog.Info("Starting day18")
	s := new(solver).prep(inp)

	start := time.Now()
	got1 := s.part1()
	slog.Info("Part 1", "time", time.Since(start).String(), "Ans", got1)

	start2 := time.Now()
	got2 := s.part2()
	slog.Info("Part 2", "time", time.Since(start2).String(), "Ans", got2)
}

type solver struct {
	patterns []string
	designs  []string
}

func (s *solver) prep(inp string) *solver {
	rows := strings.Split(inp, "\n\n")
	s.patterns = strings.Split(rows[0], ", ")
	s.designs = strings.Split(rows[1], "\n")
	return s
}

func filterPatterns(design string, patterns []string) []string {
	out := make([]string, 0, len(patterns))

	for i := range patterns {
		if strings.Contains(design, patterns[i]) {
			out = append(out, patterns[i])
		}
	}

	return out
}

func (s *solver) part1() int {
	res := 0
	cache := make(map[string]int, len(s.patterns))

	for i := range s.designs {
		designPatterns := filterPatterns(s.designs[i], s.patterns)
		if val := countVariations(cache, s.designs[i], designPatterns); val > 0 {
			res++
		}
	}

	return res
}

func (s *solver) part2() int {
	res := 0
	cache := make(map[string]int, len(s.patterns))

	for i := range s.designs {
		designPatterns := filterPatterns(s.designs[i], s.patterns)
		res += countVariations(cache, s.designs[i], designPatterns)
	}

	return res
}

func countVariations(cache map[string]int, design string, patterns []string) int {
	if len(design) == 0 {
		return 1
	}

	if val, ok := cache[design]; ok {
		return val
	}

	opts := make([]string, 0, len(patterns))
	for i := range patterns {
		if strings.HasPrefix(design, patterns[i]) {
			opts = append(opts, patterns[i])
		}
	}

	if len(opts) == 0 {
		return 0
	}

	for i := range opts {
		cache[design] += countVariations(cache, design[len(opts[i]):], patterns)
	}

	return cache[design]
}

File: day19/src/main.rs
use pkg::measure_time;
use std::collections::HashMap;
use std::fs;
mod main_test;

fn main() {
    let content = fs::read_to_string("./day19/inp");
    match content {
        Ok(content) => {
            let s: Solver = Solver::new().prep(content);
            measure_time(|| s.part1());
            measure_time(|| s.part2());
        }
        Err(e) => {
            println!("{:?}", e);
        }
    }
}

struct Solver {
    patterns: Vec<String>,
    designs: Vec<String>,
}

impl Solver {
    fn new() -> Solver {
        Solver {
            patterns: vec![],
            designs: vec![],
        }
    }

    fn prep(mut self, inp: String) -> Solver {
        let lines: Vec<&str> = inp.split("\n\n").collect();

        self.patterns = lines
            .first()
            .unwrap()
            .split(", ")
            .map(|s| s.to_string())
            .collect();
        self.designs = lines
            .last()
            .unwrap()
            .lines()
            .map(|s| s.to_string())
            .collect();

        self
    }

    fn part1(&self) -> i64 {
        let mut cache: HashMap<String, i64> = HashMap::new();

        let mut res: i64 = 0;

        for i in 0..self.designs.len() {
            if count_occurances(
                &mut cache,
                &self.designs[i],
                &filter_patterns(&self.designs[i], &self.patterns),
            ) > 0
            {
                res += 1;
            }
        }

        res
    }

    fn part2(&self) -> i64 {
        let mut cache: HashMap<String, i64> = HashMap::new();

        let mut res: i64 = 0;

        for i in 0..self.designs.len() {
            res += count_occurances(
                &mut cache,
                &self.designs[i],
                &filter_patterns(&self.designs[i], &self.patterns),
            );
        }

        res
    }
}

fn filter_patterns(design: &String, patterns: &[String]) -> Vec<String> {
    let mut out: Vec<String> = Vec::new();

    for p in patterns {
        if design.contains(p) {
            out.push(p.to_string());
        }
    }

    out
}

fn count_occurances(cache: &mut HashMap<String, i64>, design: &String, patterns: &[String]) -> i64 {
    if design.is_empty() {
        return 1;
    }

    if let Some(val) = cache.get(design) {
        return *val;
    }

    let mut res: i64 = 0;
    for pt in patterns.iter() {
        if design.starts_with(pt) {
            res += count_occurances(cache, &design[pt.len()..].to_string(), patterns);
        }
    }

    cache.insert(design.clone(), res);

    return res;
}
